POEM_081.md

```diff
@@ -0,0 +1,404 @@
+        The driver for the problem. If not specified, a simple "Run Once" driver will be used.
+    comm : MPI.Comm or <FakeComm> or None
+        The global communicator.
+    name : str or None
+        Problem name. Can be used to specify a Problem instance when multiple Problems
+        exist.
+    reports : str, bool, None, _UNDEFINED
+        If _UNDEFINED, the OPENMDAO_REPORTS variable is used. Defaults to _UNDEFINED.
+        If given, reports overrides OPENMDAO_REPORTS. If boolean, enable/disable all reports.
+        Since none is acceptable in the environment variable, a value of reports=None
+        is equivalent to reports=False. Otherwise, reports may be a sequence of
+        strings giving the names of the reports to run.
+    prob_options : dict or None
+        Remaining named args for problem that are converted to options.
+    inputs : list of str or tuple
+        List of desired inputs to subproblem. If an element is a str, then it should be
+        the var name in its promoted name. If it is a tuple, then the first element 
+        should be the promoted name, and the second element should be the var name
+        you wish to refer to it by within the subproblem [e.g. (prom_name, var_name)].
+    outputs : list of str or tuple
+        List of desired outputs from subproblem. If an element is a str, then it should be
+        the var name in its promoted name. If it is a tuple, then the first element 
+        should be the promoted name, and the second element should be the var name
+        you wish to refer to it by within the subproblem [e.g. (prom_name, var_name)].
+    **kwargs : named args
+        All remaining named args that become options for `SubproblemComp`.
+    """
+    def __init__(self, model, inputs, outputs, driver=None, comm=None,
+                 name=None, reports=_UNDEFINED, prob_options=None, **kwargs):
+
+        # check for driver and issue warning about its current use
+        # in subproblem
+        if driver is not None:
+            issue_warning('Driver results may not be accurate if'
+                          ' derivatives are needed. Set driver to'
+                          ' None if your subproblem isn\'t reliant on'
+                          ' a driver.')
+
+        # make `prob_options` empty dict to be passed as **options to problem 
+        # instantiation
+        if prob_options is None:
+            prob_options = {}
+
+        # call base class to set kwargs
+        super().__init__(**kwargs)
+
+        # store inputs and outputs in options
+        self.options.declare('inputs', {}, types=dict,
+                             desc='Subproblem Component inputs')
+        self.options.declare('outputs', {}, types=dict,
+                             desc='Subproblem Component outputs')
+
+        # set other variables necessary for subproblem
+        self._prev_complex_step = False
+
+        p = self._subprob = om.Problem(driver=driver, comm=comm, name=name,
```

Comment:
I don't think we should be instantiating the problem and setting it up here in the component `__init__`.  We should just store the args and do the actual instantiation and setup in the component's setup method.  That will ensure that we can correctly set the comm.

Reply:
One of the issues I have with instantiating the problem in `setup` is that I need to call `prob.final_setup` in order to access the model's input and output lists and get the meta data for the inputs and outputs. If I move the problem instantiation to `setup`, I'll have to call for the I/O meta data every time `setup` is called, which I figured is something we want to avoid. Is there no way we can ensure the comm is set correctly in `__init__`?